### DISCUSSION TITLE
Working with headless CMS, which is better to use between SSR and ISR for dynamic routes

### DISCUSSION THREAD

**Discussion by qpavy** [vercel/next.js#49941] [Help] [answered]
### Summary

Hi everyone,
I have a next.js app with dynamic routes defined by `/app/[pageslug]/page.tsx`. 
Those routes are fetching all pages from a headless Wordpress Rest Api.
When I build my next.js app, I am seing that /[pageslug] route is "server-side renders at runtime". So I am expecting my data to be updated everytime I am accessing a page with url like `/test`.

But when I update my page in wordpress and refreshing my next.js page, I am still seing the old data and not the updated one.
Here is the code I wrote to manage fetching dynamic pages from wordpress :

---
**api.ts**
```javascript
function get(path: string, cache: boolean = true) {
    const headers = { 'Content-Type': 'application/json' };
    return fetch(new URL(API_URL+path), {
        headers,
        method: 'GET',
        cache: cache ? 'default' : 'no-store'
    });
}

export async function getPageBySlug(slug: string) {
    try {
        const res = await get(`/page/${slug}`);
        return await res.json();
    } catch (error) {
        console.log(error);
    }
}
```
---
**/app/[pageslug]/page.tsx**
```javascript
export default async function Page({ params }: {
  params: { pageslug: string }
}) {

  const pageSlug = params.pageslug;
  const page = await getPageBySlug(pageSlug);

  return (
    <>
      <main>
        <h1 dangerouslySetInnerHTML={{__html: page.post?.title}} />
        <div className="container wp-post" dangerouslySetInnerHTML={{__html: page.post['rendered-content']}} />
      </main>
    </>
  )
}
```
---

I read that if I use fetch without revalidating, it will keep in cache the first value. So I can imagin this is why I am not seing the updated datas of the /test page. 
So I try to use ISR instead of SSG for this kind of dynamic routes. To do so, i add `generateStaticParams` on my page.tsx and create an next api route to revalidate my pages everytime I do an update on wordpress. this works well and I even can add a new page.
Here is how I did :

---
**/app/[pageslug]/page.tsx**
```javascript
export async function generateStaticParams() {
  const pages = await getAllPages();
  return pages.map((page: any) => ({
    pageslug: page.slug,
  }));
}
```
---
**BUT** 
if I want to delete the 'test' page on wordpress side, I can still use the `/test` url on next.js because the revalidating on demand keep the last value if there is an error.

So I am a little confuse about which method to use to manage dynamic pages from a headless CMS. Because I don't want my datas to be fetched on every page request, is there a way to revalidate on demand a SSR page ? Or if I use ISR, is it a way to "unvalidate" a page if the requesting page slug is responding with a 404 ?

Thank you =)

**Comment by devjiwonchoi** [comment]
First of all, in Dynamic Routes, ISR is recommended over SSR or SSG.

Referring from the official docs,

> [Error Handling and Revalidation](https://nextjs.org/docs/app/building-your-application/data-fetching/revalidating#error-handling-and-revalidation)
If an error is thrown while attempting to revalidate data, the last successfully generated data will continue to be served from the cache. On the next subsequent request, Next.js will retry revalidating the data.

In this case, I guess the easiest way to resolve this is
```
cache : 'no-store'
```


Another dirty but effective way I can think of is to validate while fetching data.
For example,

```
function get(path: string, cache: boolean = true) {
    const headers = { 'Content-Type': 'application/json' };
    return fetch(new URL(API_URL+path), {
        headers,
        method: 'GET',
        cache: cache ? 'default' : 'no-store'
    });
}

// revalidate by fetching again
const revalidateData = async (slug: string) => {
  return (await get(`/page/${slug}`)).json();
}

export async function getPageBySlug(slug: string) {
    try {
        const res = await get(`/page/${slug}`);
        return res.status === 200 ? await res.json() : revalidateData(slug);  // re-fetch if status isn't 200
    } catch (error) {
        console.log(error);
    }
}
```

**Reply by qpavy** [reply, marked as answer]
I found another way which is returning a notFound flag from my fetching function and call [notFound ](https://nextjs.org/docs/app/api-reference/functions/not-found) function in page.tsx if I receive the notFound flag. 
Doing so, next still respond to my deleted path but redirect to the not-found.tsx page. I think this is a good solution for performance. But I am not sure if keeping "deleted" pages in cache is a good way to do.

Exemple:

---
**api.tsx**
```javascript
export async function getPageBySlug(slug: string) {
    try {
        const res = await get(`/page/${slug}`);
        return await res.json();
    } catch (error) {
        console.log(error);
        return { notFound: true };
    }
}
```
---

**/app/[pageslug]/page.tsx**
```javascript
import { notFound } from 'next/navigation';
import { getPageBySlug } from "api";

export async function generateStaticParams() {
  const pages = await getAllPages();
  return pages.map((page: any) => ({
    pageslug: page.slug,
  }));
}

export default async function Page({ params }: {
  params: { pageslug: string }
}) {

  const pageSlug = params.pageslug;
  const page = await getPageBySlug(pageSlug);

  if (page.notFound) notFound();

  return (
    <>
      <main>
        <h1 dangerouslySetInnerHTML={{__html: page.post?.title}} />
        <div className="container wp-post" dangerouslySetInnerHTML={{__html: page.post['rendered-content']}} />
      </main>
    </>
  )
}
```

---

**Reply by devjiwonchoi** [reply]
You're right, I was dumb.. notFound() could resolve this issue in Dynamic Routes.

I would recommend you to use [revalidatePath](https://nextjs.org/docs/app/api-reference/functions/revalidatePath) and notFound together.

This could revalidate the current path and successfully remove the cache of the "deleted" page.

Example:
```
import { notFound } from 'next/navigation';
import { revalidatePath } from 'next/cache';    // import revalidatePath
import { getPageBySlug } from "api";

export async function generateStaticParams() {
  const pages = await getAllPages();
  return pages.map((page: any) => ({
    pageslug: page.slug,
  }));
}

export default async function Page({ params }: {
  params: { pageslug: string }
}) {

  const pageSlug = params.pageslug;
  const page = await getPageBySlug(pageSlug);

  if (page.notFound) {
    revalidatePath(page URL);    // insert page URL
    notFound();
  }

  return (
    <>
      <main>
        <h1 dangerouslySetInnerHTML={{__html: page.post?.title}} />
        <div className="container wp-post" dangerouslySetInnerHTML={{__html: page.post['rendered-content']}} />
      </main>
    </>
  )
}
```

**Reply by qpavy** [reply]
It looks to be a good way to do. I will try it and mark your answer as answer if it works as expected 🙏

**Reply by devjiwonchoi** [reply]
Thanks, and I hope it works :)

**Reply by qpavy** [reply]
Because the page still exist but as a not-found page, the revalidatePath don't delete it in cache. I am still able to see it even after on demand revalidation.
Maybe I need to use a middleware to handle this part of the code.

**Reply by devjiwonchoi** [reply]
I am not sure of this part,
> I am still able to see it even after on demand revalidation.

Which one do you mean by "able to see it"?

**Reply by qpavy** [reply]
I mean if I delete test article from my wordpress and try to reach `/test` with next.js it redirect me to `not-found` page but I still can see "test" page in cache:

![screenshot-1](https://github.com/vercel/next.js/assets/5684639/216ab7e3-90dc-4a24-b27d-80ca797d6ca4)

This means everytime someone is trying to reach an url that don't exist, I will have 3 new files created in cache for this route : "unexist.html", "unexist.meta" and "unexist.rsc"

**Reply by devjiwonchoi** [reply]
Doesn't that cache share throughout the whole not-found pages? Technically it is necessary in order to show the not-found page directly without fetching for "not existing" pages until it is revalidated.

**Reply by qpavy** [reply]
Yes I understand that those file are created to show the not-found page without fetching. But in this case it is not really usefull as it's a not-found page, it is not a problem if it takes a long time to load.
The problem is if a robot try many urls and those are redirecting to not found, I will have thousand of files in cache redirecting to not found =/

The best would be to say next.js to Fetch at every request if a route redirect to not-found.

I saw this [answer](https://stackoverflow.com/a/75731979/4291338) about clearing cache for not used files but I think it is a little dirty

**Reply by devjiwonchoi** [reply]
You're right. In that case, I guess the best option is to use both `notFound()` and `cache: no-store`, or just re-build after deleting numerous amount of pages.

**Reply by qpavy** [reply]
Yes you're right, I think I will launch a rebuild in a cron every night So even in case there is a large number of "not-found" page created in cache they will be deleted every 24 hours. Thank you for your help =)

**Reply by devjiwonchoi** [reply]
Just one another opinion, you could use webhooks to automate next build when pages are removed.

**Reply by qpavy** [reply]
True, I begin with gitlab build scheduled every weeks and if I have to much removed page I will swap to this option =)

**Comment by levipadre** [comment]
@qpavy I have a similar issue too. I'm building a project with a headless CMS, and all pages are generated by API, so my page path is `app/[[...slug]]/page.tsx`. 
But I can't revalidate any of the pages by path, for example `https://domain.com/api/revalidate?path=/about-us`

**api.ts**
```
function get(path: string) {
    const headers = { 'Content-Type': 'application/json' };
    return fetch(new URL(`https://domain.com/${path}`), {
        headers,
        method: 'GET'
    });
}

export async function getPages() {
    try {
        const res = await get(`/pages`);
        return await res.json();
    } catch (error) {
        console.log(error);
        return { notFound: true };
    }
}

export async function getPageByUrl(slug: string[]) {
    try {
        const path = (typeof slug === 'undefined' || slug[0] === 'index') ? '' : slug.join('/');
        const res = await get(`pages?url=/${path}`);
        return await res.json();
    } catch (error) {
        console.log(error);
        return { notFound: true };
    }
}
```

**/app/[[...slug]]/page.tsx**
```
export async function generateStaticParams() {
    const PAGES = await Promise.all([getPages()]);

    return PAGES.map((page: any) => ({
        slug: page.slug
    }));
}

export default async function Page({
    params
}: {
    params: {
        slug: string[]
    }
}) {
    const slug = params.slug;
    const [PAGE] = await Promise.all([getPageByUrl(slug)]);

    if (PAGE.notFound) notFound();

    return (
        <div className='container'>
            <h1>title: {PAGE[0].title}</h1>
            <div dangerouslySetInnerHTML={{__html:PAGE[0].body[0].section}}></div>
        </div>
    )
}
```

When I build it, it shows as SSG
```
Route (app)                                Size     First Load JS
┌ ○ /_not-found                            0 B                0 B
├ ● /[[...slug]]                           403 B          77.6 kB
├   └ /
├ λ /api/revalidate                        0 B                0 B
├ ○ /contact                               141 B          77.4 kB
├ ○ /favicon.ico                           0 B                0 B
├ λ /robots.txt                            0 B                0 B
└ λ /sitemap.xml                           0 B                0 B
+ First Load JS shared by all              77.2 kB
  ├ chunks/2443530c-205bc6bce197be04.js    50.5 kB
  ├ chunks/488-d4d62e10403044ce.js         24.7 kB
  ├ chunks/main-app-3a7939509f543115.js    216 B
  └ chunks/webpack-ec549998470bf1bb.js     1.73 kB

Route (pages)                              Size     First Load JS
─ ○ /404                                   181 B          74.8 kB
+ First Load JS shared by all              74.6 kB
  ├ chunks/framework-8883d1e9be70c3da.js   45.1 kB
  ├ chunks/main-032eefb970051999.js        27.6 kB
  ├ chunks/pages/_app-b555d5e1eab47959.js  195 B
  └ chunks/webpack-ec549998470bf1bb.js     1.73 kB

λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
○  (Static)  automatically rendered as static HTML (uses no initial props)
●  (SSG)     automatically generated as static HTML + JSON (uses getStaticProps)
```

Can you see any issues with it? Do you have any advice on why revalidation doesn't work and get the old data?

**Reply by qpavy** [reply]
What error are you getting when you call ` res.revalidate(req.query.path)` ?

**Reply by levipadre** [reply]
Sorry, but where should I add this code?

**Reply by qpavy** [reply]
When you use SSG, your pages data is generated at build time. You have to revalidate the path for the page you are creating / modifying / deleting in your CMS.
To do that, you have to create the `revalidate.ts` file in your next.js application inside `/pages/api` this create an api route which respond to the url `/api/revalidate?path=<path to revalidate>`.

---

**/pages/api/revalidate.ts**
```javascript
import { NextApiRequest, NextApiResponse } from "next";

export default async function handler(
    req: NextApiRequest,
    res: NextApiResponse
) {
    const path = req.query.path as string;

    try {
        await res.revalidate(path);
    } catch (error) {
        console.log('Error Revalidating : ', error);
        return res.status(500).send('Error revalidating')
    }

    return res.json({ revalidated: true});
}
```

---

This code in your nex.js app retrieve the path query param and call revalidate for this path. Revalidation make next.js "rebuild" the page corresponding to the path you are giving and this permit to update page data.

So when you add / update / delete content in your CMS, you have to call the revalidate url of your next.js app for all the path that need the updated data.

One issue with your generateStaticParams is that your app will create all the page at build time so the build can be really long if you have a large amount of pages. 
What you can do is return an empty array in `generateStaticParams` so Next will not build all your page at build time but only the first time someone is calling the page !

**Reply by levipadre** [reply]
Wow, it totally worked. I'm still testing, but it seems fine at first glance. Thank you so much @qpavy!
I just don't understand why **/app/api/revalidate/route.ts** didn't work the same way.
I was able to call on https://domain.com/api/revalidate?path=/about-us&secret=123456 (I also asked for the secret token) and it returned with `{"revalidated":true,"now":1685721969339}` as I read about it here: https://nextjs.org/docs/app/building-your-application/data-fetching/revalidating#using-on-demand-revalidation

My route.ts was like this:
```
import { NextRequest, NextResponse } from 'next/server';
import { revalidatePath, revalidateTag } from 'next/cache';
 
export async function GET(request: NextRequest) {
    const secret = request.nextUrl.searchParams.get('secret') as string;

    if(secret !== process.env.API_SECRET_TOKEN) {
        return NextResponse.json({ message: 'Invalid token' });
    }

    const tag = request.nextUrl.searchParams.get('tag') as string|null;

    if (tag) {
        revalidateTag(tag)
    }

    const path = request.nextUrl.searchParams.get('path') as string|null;

    if (path) {
        revalidatePath(path)
    }

    return NextResponse.json({ revalidated: true, now: Date.now() });
}
```

**Reply by qpavy** [reply]
I haven't use revalidatePath and revalidateTag yet but maybe take a look at https://github.com/vercel/next.js/issues/49778#issuecomment-1546968234 or https://github.com/vercel/next.js/issues/49387. This could answer your question about those two.
